packages/native-stack/src/views/RetainContext.tsx

```diff
@@ -0,0 +1,138 @@
+import type {
+  ParamListBase,
+  Route,
+  StackNavigationState,
+} from '@react-navigation/native';
+import * as React from 'react';
+
+import type {
+  NativeStackDescriptorMap,
+  NativeStackNavigationHelpers,
+} from '../types';
+
+type RetainContextT = {
+  retain(): string;
+  release(key: string): boolean;
+  restore(key: string): boolean;
+  supported: boolean;
+};
+
+/**
+ * This context allows you to retain screens in memory, even after
+ * they have been popped off the stack. This is important for
+ * supporting native iOS "Picture in Picture" mode, as the
+ * originating UIViewController must be retained in memory for the
+ * PiP overlay to continue running.
+ * Retained screens can also be "restored" to the top of the stack,
+ * for example when the user taps the button in the PiP overlay
+ * to return to the app.
+ */
+export const RetainContext = React.createContext<RetainContextT>({
+  retain() {
+    throw new Error(`Not in a native screen stack`);
+  },
+  release() {
+    throw new Error(`Not in a native screen stack`);
+  },
+  restore() {
+    throw new Error(`Not in a native screen stack`);
+  },
+  supported: false,
+});
+
+type RetainedScenes = {
+  [key: string]: {
+    route: Route<any>;
+    descriptor: NativeStackDescriptorMap['key'];
+  };
+};
+
+export function useRetainContext(
+  state: StackNavigationState<ParamListBase>,
+  navigation: NativeStackNavigationHelpers,
+  descriptors: NativeStackDescriptorMap
+) {
+  const [retainedScenes, setRetainedScenes] = React.useState<RetainedScenes>(
+    {}
+  );
+  const latestValues = React.useRef({ state, retainedScenes, descriptors });
+  React.useEffect(() => {
+    latestValues.current = { state, retainedScenes, descriptors };
+  });
```

Comment:
I'm rather against patters like that. The new state should be the result of props and the previous state as react may batch those updates, we should not rely on what is currently in `useRef`.

Reply:
I believe this is not very needed here and it's possible to refactor the code without that.

Reply:
I assume the intention is to memoize the `retain`, `release`, `restore` functions. In that case define those functions separately with `useLatestCallback` and then they can be safely added to dependency array for `useMemo`.

With that it's unnecessary to do such a ref.